src/clj/puppetlabs/puppetserver/ring/middleware/params.clj

```diff
@@ -0,0 +1,93 @@
+;; a type that is compatible with Clojure's IOFactory, which forces us to read
+;; the request body into memory twice for requests that we have to pass down
+;; into the JRuby layer.  (Technically, the Ring specification states that the
+;; body must be an InputStream, so the maintainer of the upstream library was
+;; reluctant to accept any sort of upstream PR to work around this issue.)
+;;
+;; All of this code is copied from the upstream library, and there is just
+;; one very slight modification (see comment in `assoc-form-params` function)
+;; that allows us to avoid reading the body into memory twice.
+;;
+;; In the future, if we can handle the query parameter parsing strictly on the
+;; Clojure side and remove that code from the Ruby side, we should be able to
+;; get rid of this.  That will be much easier to consider doing once we're able
+;; to get rid of the Rack/Webrick support.
+;;
+;; If that happens, we should delete this namespace :)
+;;
+
+(defn parse-params [params encoding]
+  (let [params (codec/form-decode params encoding)]
+    (if (map? params) params {})))
+
+(defn content-type
+  "Return the content-type of the request, or nil if no content-type is set."
+  [request]
+  ;; NOTE: in the latest version of ring-core, they only look in
+  ;;  the headers map for the content type.  They no longer fall
+  ;;  back to looking for it in the main request map.
+  (if-let [type (or (get-in request [:headers "content-type"])
+                    (get request :content-type))]
+    (second (re-find #"^(.*?)(?:;|$)" type))))
+
+(defn urlencoded-form?
+  "True if a request contains a urlencoded form in the body."
+  [request]
+  (if-let [^String type (content-type request)]
+    (.startsWith type "application/x-www-form-urlencoded")))
+
+(defn assoc-query-params
+  "Parse and assoc parameters from the query string with the request."
+  [request encoding]
+  (merge-with merge request
+              (if-let [query-string (:query-string request)]
+                (let [params (parse-params query-string encoding)]
+                  {:query-params params, :params params})
+                {:query-params {}, :params {}})))
+
+(defn assoc-form-params
+  "Parse and assoc parameters from the request body with the request."
+  [request encoding]
+  (merge-with merge request
+              (if-let [body (and (urlencoded-form? request) (:body request))]
+                (let [params (parse-params
+                               ;; NOTE: this is the main difference between our
+                               ;;  copy of this code and the upstream version:
+                               ;;  the upstream always does a slurp here, while
+                               ;;  we only do a slurp if the body is not already
+                               ;;  a string.
+                               (if (string? body)
+                                 body
+                                 (slurp body :encoding encoding))
```

Comment:
If we go down this path in Puppet Server code, `body` will be a `string`.  We will have pre-slurped it from the `InputStream` handed down from the webserver for the request - see https://github.com/puppetlabs/puppet-server/blob/puppet-server-1.0.8/src/clj/puppetlabs/services/request_handler/request_handler_core.clj#L86.  I'm thinking the else clause is probably just defensive programming for a hypothetical future case where it could still be passed in as an `InputStream` - like maybe once we disprove the need to pre-slurp it for preservation as a string when passed down to core Ruby Puppet code.